### DISCUSSION TITLE
[BUG]

### DISCUSSION THREAD

**Discussion by GraemeHenderson** [pnp/powershell#4103] [General]
### Notice
Many bugs reported are actually related to the PnP Framework which is used behind the scenes. Consider carefully where to report an issue:

1. **Are you using ```Invoke-PnPSiteTemplate``` or ```Get-PnPSiteTemplate```**? The issue is most likely related to the Provisioning Engine. The Provisioning engine is _not_ located in the PowerShell repo. Please report the issue here: https://github.com/pnp/pnpframework/issues.
1. **Is the issue related to the cmdlet itself, its parameters, the syntax, or do you suspect it is the code of the cmdlet that is causing the issue?** Then please continue reporting the issue in this repo.
1. **If you think that the functionality might be related to the underlying libraries that the cmdlet is calling** (We realize that might be difficult to determine), please first double check the code of the cmdlet, which can be found here: https://github.com/pnp/powershell/tree/master/src/Commands. If related to the cmdlet, continue reporting the issue here, otherwise report the issue at https://github.com/pnp/pnpframework/issues

### Reporting an Issue or Missing Feature
Issue relating to running a Powershell script to remove versions of files in Sharepoint. The script times out after 40-60 minutes. We have multiple Sharepoint libraries and for example one library has 175,000 files and the script times out having done circa 10,000 files. 
### Expected behavior 
That the script does not time out. We require it to fully complete.
### Actual behavior 
 Please describe what you see instead. Please provide samples of output or screenshots.
 The script times out with the error: Error Cleaning up Version History! Exception calling "ExecuteQuery" with "0" argument(s): "The operation has timed out."

### Steps to reproduce behavior
Please include complete script or code samples in-line or linked from [gists](https://gist.github.com/)

#Config Parameters

$SiteURL = "https://abcdefg.sharepoint.com/sites/tuvwxyz"

$VersionsToKeep = 50

Try {

    #Connect to PnP Online

    Connect-PnPOnline -Url $SiteURL -Interactive

    #Get the Context

    $Ctx= Get-PnPContext

    

 

    #Get document library

    $DocumentLibrary = "Shared Documents"

    

    Write-host "Processing Document Library:"$Library.Title -f Magenta

    #Get All Items from the List - Exclude 'Folder' List Items

    $ListItems = Get-PnPListItem -List $DocumentLibrary -PageSize 2000 | Where {$_.FileSystemObjectType -eq "File"}

    #Loop through each file

    ForEach ($Item in $ListItems)

    {

        #Get File Versions

        $File = $Item.File

        $Versions = $File.Versions

        $Ctx.Load($File)

        $Ctx.Load($Versions)

        $Ctx.ExecuteQuery()

 

        Write-host -f Yellow "`tScanning File:"$File.Name

        $VersionsCount = $Versions.Count

        $VersionsToDelete = $VersionsCount - $VersionsToKeep

        If($VersionsToDelete -gt 0)

        {

            write-host -f Cyan "`t Total Number of Versions of the File:" $VersionsCount

            $VersionCounter= 0

            #Delete versions

            For($i=0; $i -lt $VersionsToDelete; $i++)

            {

                If($Versions[$VersionCounter].IsCurrentVersion)

                {

                    $VersionCounter++

                    Write-host -f Magenta "`t`t Retaining Current Major Version:"$Versions[$VersionCounter].VersionLabel

                    Continue

                }

                Write-host -f Cyan "`t Deleting Version:" $Versions[$VersionCounter].VersionLabel

                $Versions[$VersionCounter].DeleteObject()

            }

            $Ctx.ExecuteQuery()

            Write-Host -f Green "`t Version History is cleaned for the File:"$File.Name

        }

    }

   

}

Catch {

    write-host -f Red "Error Cleaning up Version History!" $_.Exception.Message

}

### What is the version of the Cmdlet module you are running?
(you can retrieve this by executing ```Get-Module -Name "PnP.PowerShell" -ListAvailable```)
Version 2.4.0

### Which operating system/environment are you running PnP PowerShell on?
- [X] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ ] Azure Functions
- [ ] Other : please specify

**Comment by jackpoz** [comment]
Please note that `$ctx.ExecuteQuery()` is the standard CSOM method https://learn.microsoft.com/en-us/previous-versions/office/sharepoint-csom/ee546606(v=office.15) and not related to PnP PowerShell.

In general you should use [Invoke-PnPQuery](https://pnp.github.io/powershell/cmdlets/Invoke-PnPQuery.html) instead as it has a retry mechanism built to handle these cases of timeouts, where retries are needed.

**Comment by GraemeHenderson** [comment]
We replaced this in the script - at line 54 but it still timed out. Do we need to also do it at line 31? Or do we need to do it only at line 31?
